kong/pdk/client.lua

```diff
@@ -241,6 +242,50 @@ local function new(self)
     return is_tls and "tls" or "tcp"
   end
 
+  ---
+  -- Returns a table of `authenticated_groups` of the currently authenticated
+  -- consumer.
+  -- @function kong.client.get_authenticated_groups
+  -- @phases access, header_filter, body_filter, log
+  -- @treturn table,nil|err
+  -- The table will have an array part to iterate over, and a hash part where each
+  -- group name is indexed by itself. Eg.
+  -- {
+  --   [1] = "users",
+  --   [2] = "admins",
+  --   users = "users",
+  --   admins = "admins",
+  -- }
+  -- @usage
+  -- local groups, err = kong.client.get_authenticated_groups()
+  -- if not err then
+  --   for i = 1, #groups_to_check do
+  --     if groups[i] == 'group1' then
+  --       return true
+  --     end
+  --   end
+  -- end
+  function _CLIENT.get_authenticated_groups()
+    check_phase(AUTH_AND_LATER)
+
+    local authenticated_groups = ngx.ctx.authenticated_groups
+    if authenticated_groups == nil then
+      return nil
+    end
+
+    if type(authenticated_groups) ~= "table" then
+      return nil, "invalid authenticated_groups, a table was expected"
+    end
+
+    local groups = {}
+    for i = 1, #authenticated_groups do
+      groups[i] = authenticated_groups[i]
+      groups[authenticated_groups[i]] = authenticated_groups[i]
+    end
```

Comment:
this only copies the array part, not the hash table part. So it doesn't match the doc comments above.

More over by returning a copy of the table we will not be. able to cache anything by the table. So any code using this will have to iterate the table on every request, without cache option.

Reply:
> this only copies the array part, not the hash table part. So it doesn't match the doc comments above.

This part was taken from the acl plugin, which creates both the array part, and the hash part. Test output shows this but maybe I miss your point:
https://github.com/Kong/kong/pull/4932/files#diff-ad1fe51fe0f72502d5325b508826376dR35

>More over by returning a copy of the table we will not be. able to cache anything by the table. So any code using this will have to iterate the table on every request, without cache option.

Yes this is true, and for the acl plugin nothing changes (i.e. cache is used by acl in other functions, should it also be used here?). I see canary plugin uses more of the acl plugin functionality than I am needing for https://github.com/Kong/kong-plugin-session/pull/13 so wondering if we expand the features to support canary, if we would use `client` or a new pdk module `groups`. The scope of this PR was to simply get the groups from the context. If we have path for where groups can live, I can use this PR to transplant the acl groups module.

Reply:
> This part was taken from the acl plugin, which creates both the array part, and the hash part. Test output shows this but maybe I miss your point:

Beats me?? you're doing a numerical loop in the code, so that does not copy the hash entries....

Reply:
duh, sorry for the noise, you create a second entry by name.... i was just looking for `pairs`. 

my bad !  😞 🔫

Reply:
> Yes this is true, and for the acl plugin nothing changes (i.e. cache is used by acl in other functions, should it also be used here?). I see canary plugin uses more of the acl plugin functionality than I am needing for Kong/kong-plugin-session#13 so wondering if we expand the features to support canary, if we would use client or a new pdk module groups.

This touches on my general comment; the whole `groups.lua` from ACL and Canary, should be moved into the PDK. That was the reason at the time to refactor the code out of ACL into a separate file that could be reused by Canary. (final step of removing the copy and have canary use the file included with ACL never was completed)

So imo all of that functionality should go into the PDK. Maybe an argument can be added to instruct to return the original table? eg.

```lua
function _CLIENT.get_authenticated_groups(no_copy)
```

Such that it still can be used for caching purposes if needed. Maybe @thibaultcha or @hishamhm have some ideas regarding the design of the API?

Ideally also the DAO entities would be moved into core imo.

Reply:
> Ideally also the DAO entities would be moved into core imo.

Yes. In fact, there is on-going discussions around the "Consumer Groups" entity (there has been many mentions of it in the past already, even as early as 2017 when designing the Services/Routes model). I would suggest we hold off this PR until the entity has been implemented in Core (which may influence how the ACL plugin behaves...).